lib/trento/support/event.ex

```diff
@@ -5,9 +5,31 @@ defmodule Trento.Event do
 
   import Trento.Type, only: [deftype: 1]
 
-  defmacro defevent(block) do
+  defmacro defevent(opts \\ [], do: block) do
     quote do
-      deftype(unquote(block))
+      @version Keyword.get(unquote(opts), :version, 1)
+
+      deftype do
+        field :version, :integer, default: @version
+        unquote(block)
+      end
+
+      def upcast(params, metadata) do
+        params
+        |> Map.put_new("version", 1)
+        |> upcast_params(metadata)
+      end
+
+      defp upcast_params(%{"version" => @version} = params, _) do
+        params
+      end
+
+      defp upcast_params(%{"version" => version} = params, metadata) do
+        params
+        |> upcast(metadata, version + 1)
```

Comment:
Where is this `upcast` with 3 params defined?

Reply:
you need to define it for every event if the upcast is needed